sdks/java/core/src/main/java/org/apache/beam/sdk/io/WholeFileIO.java

```diff
@@ -0,0 +1,241 @@
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.PDone;
+
+/**
+ * WholeFileIO.
+ */
+public class WholeFileIO {
+
+  public static Read read() {
+    return new AutoValue_WholeFileIO_Read.Builder().build();
+  }
+
+  public static Write write() {
+    return new AutoValue_WholeFileIO_Write.Builder().build();
+  }
+
+  /**
+   * Implements read().
+   */
+  @AutoValue
+  public abstract static class Read extends PTransform<PBegin, PCollection<KV<String, byte[]>>> {
+    @Nullable
+    abstract ValueProvider<String> getFilePattern();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setFilePattern(ValueProvider<String> filePattern);
+
+      abstract Read build();
+    }
+
+    public Read from(String filePattern) {
+      checkNotNull(filePattern, "FilePattern cannot be empty.");
+      return from(ValueProvider.StaticValueProvider.of(filePattern));
+    }
+
+    public Read from(ValueProvider<String> filePattern) {
+      checkNotNull(filePattern, "FilePattern cannot be empty.");
+      return toBuilder().setFilePattern(filePattern).build();
+    }
+
+    @Override
+    public PCollection<KV<String, byte[]>> expand(PBegin input) {
+      checkNotNull(
+          getFilePattern(),
+          "Need to set the filePattern of a WholeFileIO.Read transform."
+      );
+
+      String filePattern = getFilePattern().get();
+
+      PCollection<String> filePatternPCollection = input.apply(Create.of(filePattern));
+
+      PCollection<ResourceId> resourceIds = filePatternPCollection.apply(
+          ParDo.of(
+              new DoFn<String, ResourceId>() {
+                @ProcessElement
+                public void processElement(ProcessContext c) {
+                  String filePattern = c.element();
+                  try {
+                    List<MatchResult> matchResults = FileSystems.match(
+                        Collections.singletonList(filePattern));
+                    for (MatchResult matchResult : matchResults) {
+                      List<MatchResult.Metadata> metadataList = matchResult.metadata();
+                      for (MatchResult.Metadata metadata : metadataList) {
+                        ResourceId resourceId = metadata.resourceId();
+                        c.output(resourceId);
+                      }
+                    }
+                  } catch (IOException e) {
+                    e.printStackTrace();
+                  }
+                }
+              }
+          )
+      );
+
+      PCollection<KV<String, byte[]>> files = resourceIds.apply(
+          ParDo.of(
+              new DoFn<ResourceId, KV<String, byte[]>>() {
+                @ProcessElement
+                public void processElement(ProcessContext c) {
+                  ResourceId resourceId = c.element();
+
+                  try {
+                    ReadableByteChannel channel = FileSystems.open(resourceId);
```

Comment:
* Channel needs to be closed - use try-with-resources
* There's a utility that does this - StreamUtils.getBytes(channel.newInputStream())